sdk/iot/hub/src/az_iot_hub_client_sas.c

```diff
@@ -14,104 +14,136 @@
+  AZ_RETURN_IF_FAILED(az_span_append_u32toa(signature, token_expiration_epoch_time, &signature));
 
-  AZ_RETURN_IF_FAILED(az_span_append_i32toa(*out_document, expiry_time_secs, out_document));
+  *out_signature = signature;
 
   return AZ_OK;
 }
 
-az_result az_iot_sas_token_generate(
-    az_span iothub_fqdn,
-    az_span device_id,
-    az_span signature,
-    int32_t expiry_time_secs,
+AZ_NODISCARD az_result az_iot_hub_client_sas_password_get(
+    az_iot_hub_client const* client,
+    az_span base64_hmac_sha256_signature,
+    uint32_t token_expiration_epoch_time,
     az_span key_name,
-    az_span sas_token,
-    az_span* out_sas_token)
+    az_span mqtt_password,
+    az_span* out_mqtt_password)
 {
-  AZ_PRECONDITION_VALID_SPAN(device_id, 1, false);
-  AZ_PRECONDITION_VALID_SPAN(iothub_fqdn, 1, false);
-  AZ_PRECONDITION_VALID_SPAN(signature, 1, false);
-  AZ_PRECONDITION(expiry_time_secs > 0);
-  AZ_PRECONDITION_VALID_SPAN(sas_token, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_sas_token);
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(base64_hmac_sha256_signature, 1, false);
+  AZ_PRECONDITION(token_expiration_epoch_time > 0);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_password, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_password);
 
   // Concatenates: "SharedAccessSignature sr=" scope "&sig=" sig  "&se=" expiration_time_secs
   //               plus, if key_name != NULL, "&skn=" key_name
 
   az_span sr_string = AZ_SPAN_FROM_STR(SAS_TOKEN_SR);
-  az_span devices_string = AZ_SPAN_FROM_STR(SCOPE_DEVICES_STRING);
   az_span sig_string = AZ_SPAN_FROM_STR(SAS_TOKEN_SIG);
   az_span se_string = AZ_SPAN_FROM_STR(SAS_TOKEN_SE);
   int32_t required_length = az_span_length(sr_string) + az_span_length(devices_string)
-      + az_span_length(sig_string) + az_span_length(se_string) + az_span_length(iothub_fqdn)
-      + az_span_length(device_id) + az_span_length(signature) + 5;
+      + az_span_length(sig_string) + az_span_length(se_string)
+      + az_span_length(client->_internal.iot_hub_hostname)
+      + az_span_length(client->_internal.device_id) + az_span_length(base64_hmac_sha256_signature)
+      + 5; // Number of ampersands and equal signs.
 
-  AZ_RETURN_IF_NOT_ENOUGH_CAPACITY(sas_token, required_length);
+  if (az_span_length(client->_internal.options.module_id) > 0)
+  {
+    required_length
+        += az_span_length(client->_internal.options.module_id) + az_span_length(modules_string);
+  }
 
-  *out_sas_token = sas_token;
+  AZ_RETURN_IF_NOT_ENOUGH_CAPACITY(mqtt_password, required_length);
 
   // SharedAccessSignature
-  *out_sas_token = az_span_append(*out_sas_token, sr_string);
-  *out_sas_token = az_span_append_uint8(*out_sas_token, EQUAL_SIGN);
-  *out_sas_token = az_span_append(*out_sas_token, iothub_fqdn);
-  *out_sas_token = az_span_append(*out_sas_token, devices_string);
-  *out_sas_token = az_span_append(*out_sas_token, device_id);
+  mqtt_password = az_span_append(mqtt_password, sr_string);
```

Comment:
> mqtt_password = az_span_append(mqtt_password, sr_string); [](start = 1, length = 58)

how does append return error? or it can never?
is span_append an internal api that requires capacity of source to be checked before?